src/plugins/mf_priority.cpp

```diff
@@ -61,19 +69,12 @@ int64_t priority_calculation (flux_plugin_t *p,
     if (urgency == FLUX_JOB_URGENCY_EXPEDITE)
         return FLUX_JOB_PRIORITY_MAX;
 
-    // search element in map of maps by key
-    it = users.find (userid);
+    bank = static_cast<char *> (flux_jobtap_job_aux_get (
+                                                    p,
+                                                    FLUX_JOBTAP_CURRENT_JOB,
+                                                    "mf_priority:bank"));
 
-    // search for correct fshare value using passed-in bank; otherwise, use
-    // a default bank
-    if (bank != NULL) {
-        inner_it = it->second.find (bank);
-        fshare_factor = inner_it->second;
-    }
-    else {
-        inner_it = it->second.find ("default");
-        fshare_factor = inner_it->second;
-    }
+    fshare_factor = users[userid][bank].fairshare;
```

Comment:
This is where I'm out of my depth: Should this be wrapped in a try/except block? Will the program abort if an exception is uncaught?

Reply:
I'm not sure if it is even possible, but it would be good to test that failure is handled gracefully here if an invalid bank or userid is used at this point, so that either `users[userid]` or `users[userid][bank]` do not exist. (No problem though if that would be awkward to test)

Reply:
Yeah, if there is no entry for `.fairshare` it will throw an `out_of_range` exception, so I should probably wrap this in a `try/except` block. 

I'm not sure the best strategy is to reject the job at this point; what do you think? Is there a more appropriate way to return an error and stop the job from running other than returning `flux_jobtap_reject_job ()`? The reason I ask is if there is in fact no `.fairshare` value, then the priority calculation will be incorrect.

Reply:
Yeah, I think you would have to actually raise an exception on the job, `flux_jobtap_raise_exception()`, since it is too late to call `flux_jobtap_reject_job ()`.

You could raise an non sev 0 exception, and set the priority to 0 to put the job on hold? Then if the problem if later fixed the job does not have to resubmitted?

Reply:
I think that makes sense. If the exception is a non sev 0 exception, should the `severity` be set to a positive integer? Is there a typical integer used when a value is not available? Something like this:

```c++
flux_jobtap_raise_exception (p, FLUX_JOBTAP_CURRENT_JOB, "plugin",
                             2, "plugin: fairshare value not available");
```

Sorry if this is a dumb question.

Reply:
I've used severity 4 for "warning" exceptions since this corresponds to syslog `LOG_WARNING`.
Here you may want to use 3 which is `LOG_ERR`? It is kind of arbitrary.

If the job is also held, you can note that in the exception. Also instead using `plugin: ` for the prefix, I'd suggest `mf_priority: ` so users/admins know which plugin logged the exception.

Reply:
> I'm not sure if it is even possible, but it would be good to test that failure is handled gracefully here if an invalid bank or userid is used at this point, so that either users[userid] or users[userid][bank] do not exist. (No problem though if that would be awkward to test)

I think what I can do is change the way `fairshare` is accessed from the `users` map to instead use the `at ()` function, which will throw an `out_of_range` exception if either the user or the bank cannot be found:

```c++
fshare_factor = users.at (userid).at (bank).fairshare;
```

If I wrap this in a try-catch, if an exception is thrown, I can raise a an exception on the job with an error message saying that it can't find a fairshare value and just return a priority of 0.

I think I can also confirm that this behavior occurs by sending a fake payload in the sharness tests with a null fairshare key-value pair:

```
data = {"userid": str(userid), "bank": "account4", "default_bank": "account3", "fairshare": "", "max_jobs": "10"}
```

and then confirming the priority returned for a submitted job under that bank is 0.

Reply:
Clever testing strategy!

One formatting note. I know that our coding style requests a space between `function` and parens `()`. However for this C++ case of chaining function calls, I think you could drop that if you want, e.g. instead of 
```
fshare_factor = users.at (userid).at (bank).fairshare;
```

you could do

```
fshare_factor = users.at(userid).at(bank).fairshare;
```

The second form seems more readable to me, but I leave the final decision to you!

Reply:
I agree with you. The space with chaining function calls looks little funky. I'll push up a quick fix to get rid of that space. 👍